controllers/machineset_controller.go

```diff
@@ -146,20 +146,22 @@ func (r *MachineSetReconciler) reconcile(ctx context.Context, machineSet *cluste
 		return ctrl.Result{}, errors.Wrap(err, "failed to list machines")
 	}
 
-	// Cluster might be nil as some providers might not require a cluster object
-	// for machine management.
+	// Reconcile and retrieve the Cluster object.
+	if machineSet.Labels == nil {
+		machineSet.Labels = make(map[string]string)
+	}
+	machineSet.Labels[clusterv1.MachineClusterLabelName] = machineSet.Spec.ClusterName
+
 	cluster, err := util.GetClusterFromMetadata(ctx, r.Client, machineSet.ObjectMeta)
```

Comment:
Same as above, we should be able to just get the Cluster resource from the name rather than objectmeta here.

Reply:
Opened https://github.com/kubernetes-sigs/cluster-api/issues/1547